### PR TITLE
Fix undefined symbol error in version script

### LINKED ISSUE #50414: Linker script `julia.expmap.in` references symbols that may be not defined
When trying to compile julia master on debian sid using this [Make.user](https://github.com/JuliaLang/julia/files/11952606/Make.user.txt),
if the linker is LLVM LLD 16, there's a bit of a warnings in the output:
```
    LINK /builddirs/julia-dev/usr/lib/libjulia-internal-debug.so.1.11.0
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'asprintf' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'localtime_r' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'add_library_mapping' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_IO_stdin_used' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'llvmGetPassPluginInfo' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'environ' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '__progname' failed: symbol not defined
    LINK /builddirs/julia-dev/usr/lib/libjulia-internal-debug.so.1.11
    LINK /builddirs/julia-dev/usr/lib/libjulia-internal-debug.so
    LINK /builddirs/julia-dev/usr/lib/libjulia-codegen-debug.so.1.11.0
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '__stack_chk_guard' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'asprintf' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'arraylist_grow' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'small_arraylist_grow' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'small_typeof' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'rec_backtrace' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'libsupport_init' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'localtime_r' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash32' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash32_seed' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'memhash_seed' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'restore_signals' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'add_library_mapping' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'jlbacktrace' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'jlbacktracet' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_IO_stdin_used' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_Z24jl_coverage_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_Z22jl_coverage_alloc_lineN4llvm9StringRefEi' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '_Z22jl_malloc_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol 'environ' failed: symbol not defined
ld: warning: version script assignment of 'JL_LIBJULIA_1.11' to symbol '__progname' failed: symbol not defined
    LINK /builddirs/julia-dev/usr/lib/libjulia-codegen-debug.so.1.11
    LINK /builddirs/julia-dev/usr/lib/libjulia-codegen-debug.so
    JULIA /builddirs/julia-dev/usr/lib/julia/corecompiler.ji
```
... and if LLVM LLD 17 is used, those warnings are errors for some reason,
which makes building impossible. This is quite unfortunate.

cc @topolarity @vtjnash

### LINKED ISSUE #54533: Broken linkerscript, julia fails to link with modern lld versions
Building v1.10.3 with 
```
$ ld -v
Debian LLD 18.1.6 (compatible with GNU linkers)
```
fails:
```
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '__stack_chk_guard' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'asprintf' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'rec_backtrace' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'localtime_r' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'restore_signals' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'add_library_mapping' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'jlbacktrace' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'jlbacktracet' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '_Z24jl_coverage_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '_Z22jl_coverage_alloc_lineN4llvm9StringRefEi' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '_Z22jl_malloc_data_pointerN4llvm9StringRefEi' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'llvmGetPassPluginInfo' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol 'environ' failed: symbol not defined
ld: error: version script assignment of 'JL_LIBJULIA_1.10' to symbol '__progname' failed: symbol not defined
```

### PR DESCRIPTION
**Pull request by Zentrik**
lld 17 and above by default error if symbols listed in the version script are undefined. Julia has a few of these, as some symbols are defined conditionally in Julia (e.g. based on OS), others perhaps have been removed from Julia and other seem to be defined in other libraries. Further the version script is used in linking three times, each time linking together different objects and so having different symbols defined.

Adding `-Wl,--undefined-version` is not a great solution as passing that to ld < 2.40 errors and there doesn't seem to be a great way to check if a linker supports this flag. 
I don't know how to get around these errors for symbols like `_IO_stdin_used` which Julia doesn't define and it seems to matter whether or not they are exported, see https://libc-alpha.sourceware.narkive.com/SevIQmU3/io-stdin-used-stripped-by-version-scripts. So I've converted all undefined symbols into wildcards to work around the error.

Fixes #50414, fixes #54533 and replaces #55319.

### CODE REVIEW

File: Make.inc [hunk 1]
```diff
@@ -1417,7 +1417,7 @@ ifeq (,$(findstring aarch64,$(ARCH)))
 OSLIBS += -lgcc_s
 endif
 
-OSLIBS += -Wl,--export-dynamic -Wl,--undefined-version -Wl,--version-script=$(BUILDROOT)/src/julia.expmap \
+OSLIBS += -Wl,--export-dynamic -Wl,--version-script=$(BUILDROOT)/src/julia.expmap \
```
Comment (ararslan) [inline Make.inc:1420]: I added `-Wl,--undefined-version` here as it was needed for FreeBSD AArch64. I'll build this PR on that system and see if your changes also fix the issue I hit.

Comment (ararslan) [reply Make.inc:1420]: Yep seems to work

File: src/julia.expmap.in [hunk 2]
```diff
@@ -2,41 +2,41 @@
...
     utf8proc_*;
-    jlbacktrace;
-    jlbacktracet;
-    _IO_stdin_used;
-    _Z24jl_coverage_data_pointerN4llvm9StringRefEi;
-    _Z22jl_coverage_alloc_lineN4llvm9StringRefEi;
-    _Z22jl_malloc_data_pointerN4llvm9StringRefEi;
+    jlbacktrace*;
+    jlbacktracet*;
+    _IO_stdin_used*;
```
Comment (fingolfin) [inline src/julia.expmap.in:28]: Maybe add a comment explaining the relevance?
```suggestion
    _IO_stdin_used*;  /* needed as GNU libc checks for this, see https://libc-alpha.sourceware.narkive.com/SevIQmU3/io-stdin-used-stripped-by-version-scripts */
```

**Comment by fingolfin** [comment]
Did you simply add `*` to *all* symbols? Or just those that caused you problems?

I wonder if applying this too widely might hide some issues? I.e. if a symbol truly is gone, and we just forgot to remove it from the version script then it might be good if on some platform we get an error, so that we can remove it (though it would be better if the error just showed on all platforms, otherwise "fringe" platforms suffer unfairly from "bugs" accidentally introduced by devs on the most active platforms).

That said, overall I'd like to approach these things pragmatically: this seems to resolve a real issue for people and I don't see any major downside, so I'd just apply it (and if somebody has a better idea: they are welcome to submit a PR improving things further).

**Review by fingolfin — APPROVED** [review]

**Comment by Zentrik** [comment]
I did add `*` to all symbols but iirc it's essentially required for all of them as libjulia-codegen doesn't use most of them, see codegen.expmap in https://github.com/JuliaLang/julia/pull/55319/files for a minimal expmap where I didn't add any * (only `llvmGetPassPluginInfo` gets used but that isn't used when linking flisp anyways).

**Comment by giordano** [comment]
Can you please rebase?  There's a conflict in the expmap 🥲

File: src/julia.expmap.in [hunk 3]
```diff
@@ -25,7 +25,7 @@
     utf8proc_*;
     jlbacktrace*;
     jlbacktracet*;
-    _IO_stdin_used*;
+    _IO_stdin_used*; /* See https://git.kongjian.xin:8442/FGCOM/asterisk/commit/80edcdcc45c0ed6c9223af59b2c6961df2e7a325 for why this is necessary */
```
[the remark this answers is not in the document]
Comment (Zentrik) [reply src/julia.expmap.in]: I had added this link to the wayback machine, but replaced with a bit more explanation and hopefully a more stable link.

**Pull request merged**